### ISSUE TITLE
How to implement DiffTraj-wo/Con ?

### ISSUE THREAD

**Issue by erik-buchholz** [Yasoz/DiffTraj#11]
Making this a separate issue:

How to implement the `DiffTraj-wo/Con` from Table 1 in the paper?

If I would like to run the model without the conditional guidance, i.e., DiffTraj-wo/Con in your paper, I should use the Model instead of Guide_UNet and set the `extra_embed` to none when calling it. Is this correct? Or is there anything else that needs to be done? In the config, I could change 'attr_dim': to 0 in this case?
So, what would the code for DiffTraj-wo/Con have to look like?

Like this?
```python
class Uncond_UNet(nn.Module):
    """
    This is the model without conditional information: DiffTraj-wo/Con
    """
    def __init__(self, config):
        super(Uncond_UNet, self).__init__()
        self.config = config
        self.unet = Model(config)

    def forward(self, x, t):
        pred_noise = self.unet(x, t, extra_embed=None)
        return pred_noise
```

Or did you still include the place embedding as it does not actually make use of the attributes?

I.e., like this:

```python
class Uncond_UNet(nn.Module):
    """
    This is the model without conditional information: DiffTraj-wo/Con
    """
    def __init__(self, config):
        super(Uncond_UNet, self).__init__()
        self.config = config
        self.ch = config.model.ch * 4
        self.attr_dim = config.model.attr_dim
        self.guidance_scale = config.model.guidance_scale
        self.unet = Model(config)

        # Remove guide embedding but keep place embedding
        # self.guide_emb = WideAndDeep(self.ch)
        self.place_emb = WideAndDeep(self.ch)

    def forward(self, x, t, attr=None):
        # Remove guide
        # guide_emb = self.guide_emb(attr)
        # Keep place embedding
        place_vector = torch.zeros((x.shape[0], 8), device=x.device)
        place_emb = self.place_emb(place_vector)

        # cond_noise needs to be removed b/c we don't have guide information
        # cond_noise = self.unet(x, t, guide_emb)
        uncond_noise = self.unet(x, t, place_emb)
        
        # Not sure how to define pred_noise
        # pred_noise = cond_noise + self.guidance_scale * (cond_noise -
                                                         uncond_noise)
        # Maybe like this?
        pred_noise = uncond_noise
        
        return pred_noise
```

Thanks a lot for your help!

_Originally posted by @erik-buchholz in https://github.com/Yasoz/DiffTraj/issues/10#issuecomment-2151385418_

**Comment by Yasoz** [comment]
I think both of these can be unconditional Model implementations. Since I didn't have the WideAndDeep module in my earliest implementation, I was using the first approach. But for the sake of straightforward implementation, it might be easier to just use the second one.

**Comment by erik-buchholz** [comment]
For future reference: 
I finally got DiffTraj-wo/Con to work on both of my datasets, and it performs quite well.
I ended up using the following code for the unconditional model:

```python
class Uncond_UNet(nn.Module):
    """
    This is the model without conditional information.
    """
    def __init__(self, config):
        super().__init__()
        self.config = config
        self.unet = Model(config)

    def forward(self, x, t, attr=None):
        """

        :param x: Noisy sample
        :param t: Timestep
        :param attr: Ignored, only for compatibility
        :return: Predicted noise
        """
        pred_noise = self.unet(x, t, extra_embed=None)
        return pred_noise
```

The training routine is mainly unchanged, except for ignored the `head` information.

```python
    for epoch in tqdm(range(1, n_epochs + 1), desc="Epochs", leave=True):
        for _, (trainx, head) in tqdm(enumerate(dataloader), desc="Batches", total=len(dataloader), leave=False):
            step += 1

            # Move to device
            x0 = trainx.to(device=device)
            # x0 has shape: (batch_size, 2, traj_length)
            # head = head.to(device=device)  # Ignored as we don't use head information

            # Sample the diffusion steps
            t = torch.randint(low=0, high=n_diff_steps,
                              size=(len(x0) // 2 + 1,)).to(device=device)
            t = torch.cat([t, n_diff_steps - t - 1], dim=0)[:len(x0)]
            # Get the noised images (xt) and the noise (our target)
            xt, noise = q_xt_x0(x0, t, alpha_bar=alpha_bar)
            # Run xt through the network to get its predictions
            pred_noise = model(xt.float(), t)
            # Compare the predictions with the targets
            loss = F.mse_loss(noise.float(), pred_noise)
            # Write loss to tensorboard
            writer.add_scalar("Loss/Loss", loss.item(), step)
            optim.zero_grad()
            loss.backward()
            optim.step()
            if config.model.ema:
                ema_helper.update(model)
```

I hope this might help anyone trying to run the model in the future. :)
I tested this model on Geolife and FS-NYC.

**Comment by sccaizyy-art** [comment]
> For future reference: I finally got DiffTraj-wo/Con to work on both of my datasets, and it performs quite well. I ended up using the following code for the unconditional model:
> 
> class Uncond_UNet(nn.Module):
>     """
>     This is the model without conditional information.
>     """
>     def __init__(self, config):
>         super().__init__()
>         self.config = config
>         self.unet = Model(config)
> 
>     def forward(self, x, t, attr=None):
>         """
> 
>         :param x: Noisy sample
>         :param t: Timestep
>         :param attr: Ignored, only for compatibility
>         :return: Predicted noise
>         """
>         pred_noise = self.unet(x, t, extra_embed=None)
>         return pred_noise
> The training routine is mainly unchanged, except for ignored the `head` information.
> 
>     for epoch in tqdm(range(1, n_epochs + 1), desc="Epochs", leave=True):
>         for _, (trainx, head) in tqdm(enumerate(dataloader), desc="Batches", total=len(dataloader), leave=False):
>             step += 1
> 
>             # Move to device
>             x0 = trainx.to(device=device)
>             # x0 has shape: (batch_size, 2, traj_length)
>             # head = head.to(device=device)  # Ignored as we don't use head information
> 
>             # Sample the diffusion steps
>             t = torch.randint(low=0, high=n_diff_steps,
>                               size=(len(x0) // 2 + 1,)).to(device=device)
>             t = torch.cat([t, n_diff_steps - t - 1], dim=0)[:len(x0)]
>             # Get the noised images (xt) and the noise (our target)
>             xt, noise = q_xt_x0(x0, t, alpha_bar=alpha_bar)
>             # Run xt through the network to get its predictions
>             pred_noise = model(xt.float(), t)
>             # Compare the predictions with the targets
>             loss = F.mse_loss(noise.float(), pred_noise)
>             # Write loss to tensorboard
>             writer.add_scalar("Loss/Loss", loss.item(), step)
>             optim.zero_grad()
>             loss.backward()
>             optim.step()
>             if config.model.ema:
>                 ema_helper.update(model)
> I hope this might help anyone trying to run the model in the future. :) I tested this model on Geolife and FS-NYC.

Where can I find your reproduced version of the paper "DiffTraj"? Thank you!
